### ISSUE TITLE
[Q] Usage of MemoryInputPlugin with FFMPEG input on C++ code

### ISSUE THREAD

**Issue by mekhti** [tsduck/tsduck#1087]
I can't find any usage example code, about how to send FFMPEG encoder output to the TSDUCK MemoryInputPlugin. Please, can you explain in more detail, how to use receive method of MemoryInputPlugin class?

**Comment by lelegard** [comment]
Hi @mekhti 

I assume here that everything run in one single process, your application, calling the FFmpeg encoding library and TSDuck library. 

There is an example of using the memory input and output plugins from C++ here: https://github.com/tsduck/tsduck/blob/master/sample/sample-memory-plugins/sample-memory-plugins.cpp

In the example, the input is just a few hard-coded packets. In your case, you need to add some application plumbing to extract new packets from the encoder when the input plugin sends an input event.

However, this example is just educational. I do not recommend to use the memory input/output plugins from C++. RTFM as usual and the FM says, in the description of the `memory` input plugin: _"In practice, this plugin is useful to Java and Python developers only since it is not possible to develop TSDuck plugins in these languages. To develop some custom form of transport stream input, C++ developers should rather develop their own input plugin. This is simpler and more efficient than handling input events for the memory plugin."_.

There is a sample application defining, registering and using custom plugins here: https://github.com/tsduck/tsduck/tree/master/sample/sample-app-custom

In the application, the custom plugin is a packet processing one. But the principles are the same for input plugin, just derive  from `InputPlugin` instead of `ProcessorPlugin` and implement the pure virtual method `receive`. When this method is called, you need to provide input packets, waiting as long as necessary.

If you use the `memory` input plugin, the scenario is the following:
- When `tsp` needs some packets, it calls the `receive` method of the  `memory` input plugin.
- This method signals an input event.
- Your input event handler is called and you need to provide packets.

In other words, lots of work and memory manipulation for nothing, just a method of yours that is called and needs to provide packets on demand. It is better to have your input plugin and have its `receive` method called directly. The memory + event handling stuff is just a bridge for Java and Python applications as they cannot directly register a plugin in these languages.

**Comment by lelegard** [comment]
Automatically closed after 155 days without update and "close pending" label set.
